### PR TITLE
Added test file for config/setup/startupshutdown.go

### PR DESCRIPTION
**Pull request by PatelNDipen**
The test I wrote for startupshutdown passes on my linux computer and on AppVeyor.

### CODE REVIEW

File: config/setup/startupshutdown_test.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+// The Startup function's tests are symmetrical to Shutdown tests,
+// because the Startup and Shutdown functions share virtually the
+// same functionality
+func TestStartupWithBlockingCommand(t *testing.T) {
+	var startupFuncs []func() error
+
+	startupCommand := `-startup mkdir temp_dir_for_testing
+			   -startup mkdir temp_dir_for_testing &
+			   -startup highly_unlikely_to_exist_command 
+			  `
```
Comment (mholt) [inline config/setup/startupshutdown_test.go]: What are the hyphens for at the beginning of each line?

File: config/setup/startupshutdown_test.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			t.Errorf("Test %d recieved an error of:\n%v", i, err)
+		}
+		err = os.Remove("./temp_dir_for_testing")
+		if err != nil && !test.shouldRemoveErr {
+			t.Errorf("Test %d recieved an error of:\n%v", i, err)
+		}
+
+	}
+
+	exec.Command("rm", "-r", "./temp_dir_for_testing").Start()
```
Comment (mholt) [inline config/setup/startupshutdown_test.go]: Why is this needed when you call os.Remove above on line 47?

File: config/setup/startupshutdown_test.go (hunk 1 above)
Comment (PatelNDipen) [inline config/setup/startupshutdown_test.go]: I included the hyphens to indicate that "startup" is just a flag and not part of the command. I can remove the hyphens if this style of denoting flags is inconsistent with the rest of the project.

Comment (mholt) [inline config/setup/startupshutdown_test.go]: I don't think they're needed; looks kind of funny to me. :)

File: config/setup/startupshutdown_test.go (hunk 2 above)
Comment (PatelNDipen) [inline config/setup/startupshutdown_test.go]: The "startup mkdir temp_dir_for_testing &" command is nonblocking, therefore this command will not create the "temp_dir_for_testing" directory before the os.Remove call on line 47. I expect the presence of this error on line 47 to test whether the "startup mkdir temp_dir_for_testing &" command executed in a nonblocking manner.  In order to actually remove the "temp_dir_for_testing" directory after line 47 produces an error I included the remove command on line 54. 

The extra remove statement on line 54 is not needed for the execution of the "-startup mkdir temp_dir_for_testing" command, because this command is non-blocking; therefore, the os.Remove call on line 47 will not produce an error. 

I'll add comments right after the startupCommand variable to specify whether each command is blocking and whether the command should produce an error. Also I'll add a concise comment on line 54 to explain its use.

Comment (mholt) [inline config/setup/startupshutdown_test.go]: Hmm, that makes sense. But are there any guarantees that the dir will be ready to remove by line 54? It feels kind of racey right now. You need a way to block until the command is finished, then you can remove it.

Comment (PatelNDipen) [inline config/setup/startupshutdown_test.go]: As you can see from all these commits, I was having difficulty with Windows. I found that the middleware.SplitCommandAndArgs function in middleware/commands.go  removes the backslashes from windows paths.  I called this function with an arguement of ["mkdir", "C:\Users\appveyor\AppData\Local\Temp\1\temp_dir_for_testing_startupshutdown.go"], and I received a return value of "mkdir" and "C:UsersappveyorAppDataLocalTemp1temp_dir_for_testing_startupshutdown.go."  I am not sure if this is an issue to worry about. What do you think?

Comment (mholt) [inline config/setup/startupshutdown_test.go]: Ah, it uses [shlex](https://github.com/flynn/go-shlex) which parses the command shell-style. In shell, escapes are preceded with \ so that might be why they're disappearing.

Can you try adding an if check in SplitCommandAndArgs? Something like this:

```go
if runtime.GOOS == "windows" {
    cmd = strings.Replace(cmd "\\", "\\\\")
}
```

See if escaping the backslashes like this helps...

Comment (PatelNDipen) [inline config/setup/startupshutdown_test.go]: Thanks for the help. Adding the if-statement that you provided did solve the issue on Windows. However, now there are new errors occurring, because the SplitCommandAndArgs tests are OS insensitive in that the tests expect arguments such as "arg\" to be escaped by SplitCommAndArgs on both Unix and Windows. SplitCommandAndArgs does escape "\" on Unix, but since I added your if-statement into commands.go, SplitCommandAndArgs no longer escapes the "\" on Windows. Can I delete these OS insensitive test cases from commands_test.go?

Comment (mholt) [inline config/setup/startupshutdown_test.go]: Hm, well I think we're biting off more than we can chew here. The exact commands that are executed by Startup/Shutdown are just minutia that shouldn't matter here. I think fixing the Windows-backslashes problem belongs in the [tests for SplitCommandAndArgs](https://github.com/mholt/caddy/blob/136119f8ac2d6702ba69c7808956aafbdcdbdfc4/middleware/commands_test.go#L9).

It just so happened that your PR exposed a case that isn't currently under test (apparently) with SplitCommandAndArgs.

Is there a way you can test Startup/Shutdown without using slashes in the command? (It would be acceptable to make the tests execute one command for Windows and another for *nix systems.)

Comment (PatelNDipen) [inline config/setup/startupshutdown_test.go]: Yeah, I can test Startup/Shutdown in a different way.Sorry, I am not understanding what you're trying to say. Why should I refrain from using slashes in the Startup/Shutdown tests?

Comment (mholt) [inline config/setup/startupshutdown_test.go]: Because the slashes are causing problems with SplitCommandAndArgs, so if you use slashes like this, the tests for an unrelated function (Startup/Shutdown) won't pass. The other possibility is we could fix the bug in SplitCommandAndArgs first, so it works cross-platform with slashes. Then come back and finish these tests.

Comment (PatelNDipen) [inline config/setup/startupshutdown_test.go]: Sorry, I think that I was not clear in my earlier message today. Thanks to your suggested modification to SplitCommandAndArgs, the StartupAndShutdown tests, which use slashes, pass. The only issue now is that on Windows, ' \" ' would be parsed to ' \" ' instead of the intended ' " '. I could solve this bug by checking if the command is a path to a file or directory using the os library, but a call to os.Open(command) might be expensive. So if you don't want don't think it is worth it to add all this code for Windows in the SplitCommandAndArgs function, then I can definitely change the StartupShutdown test.

Comment (mholt) [inline config/setup/startupshutdown_test.go]: I see now. Like you, I don't think the os.Open(...) is a good idea... maybe we should only use shlex if we're not on Windows, and if we're on Windows... well, parse the string some other way I guess.

Comment (PatelNDipen) [inline config/setup/startupshutdown_test.go]: I see that Karthic is trying to fix the parsing bug in Windows. I guess I'll pull the breaks on StartupAndShutdown's tests until this bug is sorted out. Is there any way in which I could contribute to solving the parsing bug or should I move onto testing other files?

Comment (mholt) [inline config/setup/startupshutdown_test.go]: I don't know, but I bet Karthic would have some ideas, if there is a way. (Of course you can find him on Slack.) Meanwhile, though, setup/browse.go and setup/bindhost.go don't have any tests at the moment. :)

Comment (Karthic-Hackintosh) [inline config/setup/startupshutdown_test.go]: Hmm, I gave it a good thought. I now have communicate my thoughts well enough.There are some fundamental problems here, Let me start,
1. Let's recollect why was `shlex.Split()` necessary at first place. It was required to parse the command and its arguments to be executed from the startup/shutdown middleware and feed it to exec.
2.Let's now go down and understand why was startup/shutdown built. If not this middleware a Linux caddy user has to open a linux terminal and execute bunch of `linux shell commands`  to bootstrap before the caddy server is started. Similarly a windows user has to open the DOS shell and issue bunch of `DOS commands` before starting caddy. The way DOS and Linux shell parse commands is totally different, There is no unified way of doing it across windows and linux, If we could then we'll be solving a bigger problem !

Comment (mholt) [inline config/setup/startupshutdown_test.go]: Yep, exactly. When I used shlex I forgot about Windows :smile:

Basically, if we're on Windows, we need to split the command and its arguments in a Windows-friendly way, and if we're not on Windows, we can use shlex. From chatting with gophers in Slack it seems that parsing Windows commands can be tricky, but I bet we can get away with a simple function that splits on whitespace except in quotes. Let me see if I can whip something together and fix the issue...

**Comment by mholt** [comment]
@PatelNDipen  I think we finally found a good way to solve that issue with the backslashes, thanks to @Makpoc's help :smiley: Will you please sync with master and flatten the commits on this once the tests are passing?

**Comment by PatelNDipen** [comment]
I was getting a nil pointer exception, because the NewTestController did include a value for the OncePerServerBlock field and startupshutdown passes uses the controller's OncePerServerBlock. I fixed the nil pointer exception by added a value for the OncePerServerBlock field in controllers.go. Also, I'm fairly new to git. Is flattening the commit necessary? My latest commit (027bf49) includes the functioning startupshutdown tests and a modified controller.go file. What should I do in terms of git commands in order to prepare the committed code for merging.

**Comment by mholt** [comment]
@PatelNDipen To flatten commits, I use this technique: http://stackoverflow.com/a/11732910/1048862

And yeah, flattening these will be helpful since most of them are just to try running tests on the CI. That way it won't clutter the commit history.
